crates/sui-bridge/src/server/handler.rs

```diff
@@ -87,18 +255,202 @@ impl BridgeRequestHandlerTrait for BridgeRequestHandler {
+            target_address: EthAddress::random().as_bytes().to_vec(),
+            token_type: TokenId::USDC as u8,
+            amount: 12345,
+        };
+
+        // TODO: remove once we don't rely on env var to get package id
+        std::env::set_var("BRIDGE_PACKAGE_ID", "0x0b");
+        init_all_struct_tags();
+
+        let mut sui_event_1 = SuiEvent::random_for_testing();
+        sui_event_1.type_ = SuiToEthTokenBridgeV1.get().unwrap().clone();
+        sui_event_1.bcs = bcs::to_bytes(&emitted_event_1).unwrap();
+        let sui_tx_digest = sui_event_1.id.tx_digest;
+        sui_client_mock.add_events_by_tx_digest(sui_tx_digest, vec![sui_event_1.clone()]);
+        sui_signer_with_cache
+            .sign((sui_tx_digest, sui_event_idx))
+            .await
+            .unwrap();
+
+        // Because the result is cached now, the verifier should not be called again.
+        // Even though we remove the `add_events_by_tx_digest` mock, we will still get the same result.
+        sui_client_mock.add_events_by_tx_digest(sui_tx_digest, vec![]);
+        sui_signer_with_cache
+            .sign((sui_tx_digest, sui_event_idx))
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_eth_signer_with_cache() {
+        let (_, kp): (_, BridgeAuthorityKeyPair) = get_key_pair();
+        let signer = Arc::new(kp);
+        let eth_mock_provider = EthMockProvider::default();
+        let contract_address = EthAddress::random();
+        let eth_client = EthClient::new_mocked(
+            eth_mock_provider.clone(),
+            HashSet::from_iter(vec![contract_address]),
+        );
+        let eth_verifier = EthActionVerifier {
+            eth_client: Arc::new(eth_client),
+        };
+        let mut eth_signer_with_cache = SignerWithCache::new(signer.clone(), eth_verifier);
+
+        // Test `get_cache_entry` creates a new entry if not exist
+        let eth_tx_hash = TxHash::random();
+        let eth_event_idx = 42;
+        assert!(eth_signer_with_cache
+            .get_testing_only((eth_tx_hash, eth_event_idx))
+            .await
+            .is_none());
+        let entry = eth_signer_with_cache
+            .get_cache_entry((eth_tx_hash, eth_event_idx))
+            .await;
+        let entry_ = eth_signer_with_cache
+            .get_testing_only((eth_tx_hash, eth_event_idx))
+            .await;
+        assert!(entry_.unwrap().lock().await.is_none());
+
+        let (_, action) = get_test_log_and_action(contract_address, eth_tx_hash, eth_event_idx);
+        let sig = BridgeAuthoritySignInfo::new(&action, &signer);
+        let signed_action = SignedBridgeAction::new_from_data_and_sig(action.clone(), sig);
+        entry.lock().await.replace(Ok(signed_action));
+        let entry_ = eth_signer_with_cache
+            .get_testing_only((eth_tx_hash, eth_event_idx))
+            .await;
+        assert!(entry_.unwrap().lock().await.is_some());
```

Comment:
Again, would be cool if you can also check the contents of the entry. Maybe include a bit complexity too for when there are multiple entries in the cache with different results, etc.